[Sink] Add a bridge for ROS support in NNStreamer and implement tensor_ros_sink using it

wooksong opened the pull request:
Addresses https://github.com/nnsuite/nnstreamer/issues/500.

This PR newly introduces a ROS package, nns_ros_bridge, that bridges between the NNStreamer and ROS frameworks. Using this package, a NNStreamer extension plug-in named tensor_ros_sink is included. At the current time, based on the tensor_sink of the NNStreamer framework, initialization of the nns_ros_bridge instance and configuration for the topic message to publish using it work.

Signed-off-by: Wook Song <wook16.song@samsung.com>

myungjoo commented on gst/tensor_ros_sink/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+	RUNTIME DESTINATION ${EXEC_PREFIX}
+	LIBRARY DESTINATION ${GST_INSTALL_DIR}
+	ARCHIVE DESTINATION ${LIB_INSTALL_DIR}
+)
```
In the final release, let's generate "libnnstreamer-ros-extension.so" including both ros-sink and ros-src.
However, until you have the full features completed, please keep go ahead with "libtensor_ros_sink.so"

wooksong replied:
@myungjoo  OK. IMO, "libnnstreamer-extension-XXX.so"  also looks pretty.

wooksong committed:
[Common/RosBridge] Add a bridge for ROS support in NNStreamer

This patch introduces nns_ros_bridge, a bridge for ROS support, to the
NNStreamer framework. In order to publish and subscribe 'tensors'
between the nns_ros_bridge-based nodes, the custom message named
'tensors' is newly defined using UInt8MultiArray of std_msgs.

Signed-off-by: Wook Song <wook16.song@samsung.com>

wooksong committed:
[Common/CMAKE] Build nns_ros_bridge using cmake only

This patch adds a CMakeLists.txt file for building the nns_ros_bridge
ROS package using cmake only instead of catkin_make.

Signed-off-by: Wook Song <wook16.song@samsung.com>

wooksong committed:
[Sink] Add tensor_ros_sink that uses nns_ros_bridge based on 'tensor_sink'

Based on 'tensor_sink' of the NNStreamer framework, this patch adds
the NNStreamer plugin, tensor_ros_sink. Using the bridge class,
initialization of the nns_ros_bridge instance, configuration for the
topic to publish, and publishing the actual topic message are working
now.

Signed-off-by: Wook Song <wook16.song@samsung.com>

wooksong committed:
[CMAKE] Add cmake files for building nnstreamer ROS extension plug-ins

This patch adds the CMakeLists.txt files for building the NNStreamer ROS
extension plug-ins.

Signed-off-by: Wook Song <wook16.song@samsung.com>

myungjoo commented on common/CMakeLists.txt:
```diff
@@ -0,0 +1,240 @@
+    SET(${return_list} ${all_list} PARENT_SCOPE)
+ENDFUNCTION(search_pkg_conf_dirs)
+
+## Set CMAKE_PREFIX_PATH to kinect default path
```
Yes, I'd like to have Kinect as well :smile:

wooksong replied:
@myungjoo  Fixed. I have no idea where 'kinect' came from :stuck_out_tongue_closed_eyes:

myungjoo commented on common/msg/tensors.msg:
```diff
@@ -0,0 +1 @@
+std_msgs/UInt8MultiArray[] tensors
```
Is this still safe and valid for tensors with float/double/int32 and so on?

wooksong replied:
std_msgs/UInt8MultiArray includes a container of vector<uint8_t>  (i.e., byte array) and a header describing what this vector actually is. In order to convert such byte vector to other type of data, the name of the actual type of published tensor is embedded in the header using tensor_element_typename[]'.

In short, IMO, this is safe way to handle all of the types of tensor but endian (ROS message is serialized with little-endian order) or data alignment issues could be addressed.

myungjoo commented on common/node/nns_ros_bridge.cc:
```diff
@@ -0,0 +1,172 @@
+  NnsRosBridge *nrb_instance = (NnsRosBridge *) instance;
+
+  return nrb_instance->setPubTopicInfo (conf);
+}
```
Check your editor.

wooksong replied:
@myungjoo  Fixed and **"files.insertFinalNewline": true** is  set to the global setting of my VS code now.

myungjoo commented on common/node/nns_ros_bridge.cc:
```diff
@@ -0,0 +1,172 @@
+ * void type pointer what the nns_ros_bridge_init() function returns).
+ */
+extern void *
+nns_ros_bridge_init (const char *node_name, const char *topic_name)
```
```extern``` ??

wooksong replied:
@myungjoo  Removed. Seems a copy-paste mistake sweat :sweat:

jaeyun-jung commented on gst/tensor_ros_sink/tensor_ros_sink.c:
```diff
@@ -0,0 +1,849 @@
+    in_tensors[i].type = self->in_config.info.info[i].type;
+  }
+
+  g_return_val_if_fail (nns_ros_bridge_publish (self->nns_ros_bind_instance,
```
need gst_memory_unmap() when failed.

wooksong replied:
@jaeyun-jung gst_memory_unmap() is added to the case that nns_ros_bridge_publish is failed.

jaeyun-jung commented on gst/tensor_ros_sink/tensor_ros_sink.c:
```diff
@@ -0,0 +1,849 @@
+      GST_RANK_NONE, GST_TYPE_TENSOR_ROS_SINK);
+}
+
+#ifndef SINGLE_BINARY
```
These lines are unnecessary now (PR #868 in nnstreamer, SINGLE_BINARY removed)

wooksong replied:
@jaeyun-jung Removed.

jaeyun-jung commented on gst/tensor_ros_sink/tensor_ros_sink.c:
```diff
@@ -0,0 +1,849 @@
+ * @brief Macro for debug message.
+ */
+#define silent_debug(...) \
+    debug_print (DBG, __VA_ARGS__)
```
Macro debug_print and err_print in tensor_common would be removed. Please change these.
nnstreamer issue 807(Revise logging mechanisms) and PR 875(Change log format)

wooksong replied:
@jaeyun-jung  Changed as well.

jaeyun-jung commented on common/node/nns_ros_bridge.cc:
```diff
@@ -0,0 +1,172 @@
+  }
+
+  this->ready_to_pub = TRUE;
+  this->nns_tensors_info_pub = (GstTensorsInfo *) &(conf->info);
```
Question : nns_tensors_info_pub is it necessary?
IMO it is better to copy the tensor info (conf->info) into nns_tensors_info_pub.

wooksong replied:
A: nns_tensors_info_pub is unnecessary now. Removed.

jaeyun-jung commented on gst/tensor_ros_sink/tensor_ros_sink.h:
```diff
@@ -0,0 +1,102 @@
+typedef struct _GstTensorRosSinkClass GstTensorRosSinkClass;
+
+/* Experimental */
+#if defined(WIN32) && (!defined(__MINGW32__))
```
I'm not sure this can work, in glib
```
#ifdef G_OS_UNIX
#include <unistd.h> /* getpid on UNIX */
#endif
#ifdef G_OS_WIN32
#include <process.h> /* getpid on win32 */
#endif
```

wooksong replied:
@jaeyun-jung Changed as follows:
```suggestion
/* Experimental: To use getpid() */
#ifdef G_OS_WIN32
#include <process.h>
#else
#include <sys/types.h>
#include <unistd.h>
#endif
```

again4you commented on common/package.xml:
```diff
@@ -0,0 +1,68 @@
+<package format="2">
+  <name>nns_ros_bridge</name>
+  <version>0.0.1</version>
+  <description>The my_first_ros_pkg package</description>
```
wrong description :)

wooksong replied:
@again4you Fixed.

wooksong committed:
[SINK] Revise debug and error logs

This patch revises debug and error logs to use GST_DEBUG/ERROR macros.

Signed-off-by: Wook Song <wook16.song@samsung.com>

wooksong commented:
### Submit v2 ###
- kinect -> kinetic
- Correct the wrong package description in package.xml for nns_ros_bridge
- Remove the meaningless extern keywords 
- Revise debug and error logs to use GST_DEBUG/ERROR macros
- Changed the trick to use getpid() in the Windows platform
- Remove nns_tensors_info_pub
- Resolve the 'No newline at end of file' issue
- Remove #ifndef for SINGLE_BINARY

myungjoo reviewed (approved):
LGTM.

myungjoo commented:
Let's Move On.
